sncast_std/src/lib.cairo

```diff
@@ -5,14 +5,99 @@ use core::fmt::{Debug, Display, Error, Formatter};
 
 #[derive(Drop, PartialEq, Serde, Debug)]
 pub struct ErrorData {
-    msg: ByteArray
+    revert_error: ContractExecutionError
 }
 
-#[derive(Drop, PartialEq, Serde, Debug)]
+#[derive(Drop, PartialEq, Debug)]
 pub struct TransactionExecutionErrorData {
     pub transaction_index: felt252,
-    // TODO(#3120): Implement `ContractExecutionError` and update below field type
-    pub execution_error: ByteArray,
+    pub execution_error: ContractExecutionError,
+}
+
+impl TransactionExecutionErrorDataSerde of Serde<TransactionExecutionErrorData> {
+    fn serialize(self: @TransactionExecutionErrorData, ref output: Array<felt252>) {
+        output.append(*self.transaction_index);
+        self.execution_error.serialize(ref output);
+    }
+    fn deserialize(ref serialized: Span<felt252>) -> Option<TransactionExecutionErrorData> {
+        let transaction_index = (*serialized.pop_front()?);
+        let execution_error = Serde::<ContractExecutionError>::deserialize(ref serialized).unwrap();
+        Option::Some(TransactionExecutionErrorData { transaction_index, execution_error })
+    }
+}
+
+
+// TODO(#3120): We should implement displaying message variant as deserialized byte array.
+// ATM, it's dispplayed as a serialized one.
+#[derive(Drop, PartialEq, Copy, Debug)]
+pub enum ContractExecutionError {
+    Nested: Box<ContractExecutionErrorInner>,
+    // We can't use `ByteArray` here, because `ContractExecutionError` must derive `Copy`.
+    // `Message` is a span of felts, which holds the serialized byte array elements.
+    Message: Span<felt252>,
+}
+
+#[derive(Drop, Serde, Copy, Debug)]
+pub struct ContractExecutionErrorInner {
+    contract_address: ContractAddress,
+    class_hash: felt252,
+    selector: felt252,
+    error: ContractExecutionError,
+}
+
+impl ContractExecutionErrorInnerPartialEq of PartialEq<Box<ContractExecutionErrorInner>> {
+    fn eq(lhs: @Box<ContractExecutionErrorInner>, rhs: @Box<ContractExecutionErrorInner>) -> bool {
+        let lhs = (*lhs).unbox();
+        let rhs = (*rhs).unbox();
+        lhs.contract_address == rhs.contract_address
+            && lhs.class_hash == rhs.class_hash
+            && lhs.selector == rhs.selector
+            && lhs.error == rhs.error
+    }
```

Comment:
Can't we derive this?

Reply:
No, we have `Box< ContractExecutionErrorInner >`, that's why.